instrumentation/spring/spring-batch-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/batch/step/TracingStepExecutionListener.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.spring.batch.step;
+
+import static io.opentelemetry.javaagent.instrumentation.spring.batch.step.StepExecutionTracer.tracer;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.javaagent.instrumentation.api.ContextStore;
+import io.opentelemetry.javaagent.instrumentation.spring.batch.ContextAndScope;
+import org.springframework.batch.core.ExitStatus;
+import org.springframework.batch.core.StepExecution;
+import org.springframework.batch.core.StepExecutionListener;
+import org.springframework.core.Ordered;
+
+public class TracingStepExecutionListener implements StepExecutionListener, Ordered {
+  private final ContextStore<StepExecution, ContextAndScope> executionContextStore;
+
+  public TracingStepExecutionListener(
+      ContextStore<StepExecution, ContextAndScope> executionContextStore) {
+    this.executionContextStore = executionContextStore;
+  }
+
+  @Override
+  public void beforeStep(StepExecution stepExecution) {
+    Context context = tracer().startSpan(stepExecution);
+    Scope scope = context.makeCurrent();
```

Comment:
same question about synchronous guarantee? for cautionary tale see https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/49c2a36811d10692f8b5b4669b150fe0fb0c4b3e/instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/TracingExecutionInterceptor.java#L87-L93